# ![logo](logo_tiny.png "nanoCEM")Output files format

## Result file from alignment_magnifier

The **alignment_feature.csv** file will be output and mismatch information will be recorded.

| Column name       |Description|
|------------|------------|
| Chrom      |transcript/chrome or gene id
| Position   |position in transcript/chrome or gene id (**1-based**)
| Base       |base of this position in reference
| Coverage   |the number of times a specific base position covered by sequencing reads
| Align string |string generated by `samtools mpileup` recording the alignment feature, please refer to their official [document](https://www.htslib.org/doc/samtools-mpileup.html)
| Q string   |string generated by `samtools mpileup` recording the base quality feature, please refer to their official [document](https://www.htslib.org/doc/samtools-mpileup.html)
| Group      |represent the group from `Sample` or `Control`. The `Sample` is the input in argument, and `Control` belongs to the control group.
| A          |the number of `A` aligned to this position
| T          |the number of `T` aligned to this position
| C          |the number of `C` aligned to this position
| G          |the number of `G` aligned to this position
| Match      |the number of bases aligned to this position that match the reference base

## Result file from current_event_magnifier

The **current_feature.csv** file will be output and current feature information will be recorded.

| Column name       |Description|
|------------|------------|
| Read ID | ID of read passed through the interest region.
| Mean |the mean value of current event of this position in this read
| STD  |the STD value of current event of this position in this read
| Median |the median value of current event of this position in this read
| Dwell time |the length of the current event of this position in this read
| Position |position in transcript/chrome or gene id (**1-based**)
| Group |represent the group from `Sample` or `Control`. The `Sample` is the input in argument, and `Control` belongs to the control group.


The **MANOVA_result.csv** file records the significant level indicating the difference between two groups

| Column name     |Description|
|-----------------|------------|
| Position        | position in transcript/chrome or gene id (**1-based**)
| P value(-log10) | P value from MANOVA applying on the PCA result. (Recommend cutoff : `2` for f5c)
| Norm difference | difference of normlized mean value of two groups
| Result          | whether it is significant after dividing according to the recommended cutoff.